h2o-core/src/main/java/water/rapids/ast/prims/time/AstMoment.java

```diff
@@ -0,0 +1,164 @@
+          timevecs.add(fr.vec(0));
+          chunksmap.add(i);
+        }
+      } else if (vi.isNum()){
+        double d = vi.getNum();
+        if (Double.isNaN(d))
+          naResult = true;
+        else
+          timeparts[i] = (int) d;
+      } else {
+        throw new IllegalArgumentException("Argument " + i + " is neither a number nor a frame");
+      }
+    }
+
+    // If all arguments are scalars, return a 1x1 frame
+    if (timevecs.isEmpty()) {
+      double val = Double.NaN;
+      if (!naResult) {
+        try {
+          val = new MutableDateTime(timeparts[0], timeparts[1], timeparts[2], timeparts[3], timeparts[4],
+                                    timeparts[5], timeparts[6], DateTimeZone.UTC).getMillis();
+        } catch (IllegalFieldValueException ignored) {}
+      }
+      return make1x1Frame(val);
+    }
+
+    // Some arguments are vecs -- create a frame of the same size
+    Vec[] vecs = timevecs.toArray(new Vec[timevecs.size()]);
+    int[] cm = ArrayUtils.unbox(chunksmap);
+    Frame fr = new SetTimeTask(timeparts, cm, naResult)
+        .doAll(Vec.T_TIME, vecs)
+        .outputFrame(Key.<Frame>make(), new String[]{"time"}, null);
+
+    return new ValFrame(fr);
+  }
+
+  private ValFrame make1x1Frame(double val) {
+    Vec v = Vec.makeTimeVec(new double[]{val}, null);
+    Frame f = new Frame(new String[]{"time"}, new Vec[]{v});
+    return new ValFrame(f);
+  }
+
+
+  private static class SetTimeTask extends MRTask<SetTimeTask> {
+    private int[] tp;
+    private int[] cm;
+    private boolean na;
+
+    /**
+     * @param timeparts is the array of [year, month, day, hrs, mins, secs, ms]
+     *                  for all constant parts of the date;
+     * @param chunksmap is a mapping between chunks indices and the timeparts
+     *                  array. For example, if {@code chunksmap = [1, 2]},
+     *                  then the first chunk describes the "month" part of the
+     *                  date, and the second chunk the "day" part.
+     */
+    public SetTimeTask(int[] timeparts, int[] chunksmap, boolean naResult) {
+      tp = timeparts;
+      cm = chunksmap;
+      na = naResult;
+    }
+
+    @Override public void map(Chunk[] chks, NewChunk nc) {
+      int nVecs = cm.length;
+      assert chks.length == nVecs;
+      MutableDateTime dt = new MutableDateTime(0, DateTimeZone.UTC);
+      int nChunkRows = chks[0]._len;
+      if (na) {
+        for (int i = 0; i < nChunkRows; i++) {
+          nc.addNum(Double.NaN);
+        }
+      } else {
+        BYROW:
+        for (int i = 0; i < nChunkRows; i++) {
+          for (int j = 0; j < nVecs; j++) {
+            double d = chks[j].atd(i);
+            if (Double.isNaN(d)) {
+              nc.addNum(Double.NaN);
+              continue BYROW;
+            }
+            tp[cm[j]] = (int) d;
+          }
+          try {
+            dt.setDateTime(tp[0], tp[1], tp[2], tp[3], tp[4], tp[5], tp[6]);
```

Comment:
You don't really need an instance of MutableDateTime, see implementation of setDateTime:

        long instant = getChronology().getDateTimeMillis(
            year, monthOfYear, dayOfMonth, hourOfDay, minuteOfHour, secondOfMinute, millisOfSecond);
        setMillis(instant);

You can just lookup the required Chronology for the UTC timezone and call getDateTimeMillis directly.